File: packages/audio-tap/android/src/main/java/com/margelo/nitro/clarion/audio_tap/HybridClarionAudioTap.kt
package com.margelo.nitro.clarion.audio_tap

import android.content.Context
import com.clarionhq.audio_tap.AudioTapCallbacks
import com.clarionhq.audio_tap.AudioTapError
import com.clarionhq.audio_tap.AudioTapSession
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.ArrayBuffer
import com.margelo.nitro.core.Promise
import java.util.concurrent.atomic.AtomicInteger

@DoNotStrip
class HybridClarionAudioTap : HybridClarionAudioTapSpec() {
  override val memorySize: Long get() = 0L

  private val context: Context = NitroModules.applicationContext
    ?: error("Nitro applicationContext is null — was the module initialized?")

  private val nextListenerId = AtomicInteger(1)
  private val frameListeners = mutableMapOf<Int, (NativeAudioTapFrame) -> Unit>()
  private val stateListeners = mutableMapOf<Int, (String) -> Unit>()
  private val statsListeners = mutableMapOf<Int, (NativeAudioTapStats) -> Unit>()
  private val errorListeners = mutableMapOf<Int, (NativeAudioTapError) -> Unit>()

  @Volatile private var session: AudioTapSession? = null
  @Volatile private var currentStateValue: String = "idle"

  private val callbacks = object : AudioTapCallbacks {
    override fun onState(state: String) {
      currentStateValue = state
      stateListeners.values.toList().forEach { runCatching { it(state) } }
    }

    override fun onFrame(
      pcm: ByteArray,
      timestamp: Double,
      frameIndex: Long,
      sampleRate: Int,
      channels: Int,
      bitsPerSample: Int,
    ) {
      val buffer = ArrayBuffer.copy(pcm)
      val frame = NativeAudioTapFrame(
        pcm = buffer,
        timestamp = timestamp,
        frameIndex = frameIndex.toDouble(),
        sampleRate = sampleRate.toDouble(),
        channels = channels.toDouble(),
        bitsPerSample = bitsPerSample.toDouble(),
      )
      frameListeners.values.toList().forEach { runCatching { it(frame) } }
    }

    override fun onStats(
      uptimeMs: Double,
      framesEmitted: Long,
      framesDropped: Long,
      bufferFillPct: Double,
    ) {
      val payload = NativeAudioTapStats(
        uptimeMs = uptimeMs,
        framesEmitted = framesEmitted.toDouble(),
        framesDropped = framesDropped.toDouble(),
        listenerCount = frameListeners.size.toDouble(),
        bufferFillPct = bufferFillPct,
      )
      statsListeners.values.toList().forEach { runCatching { it(payload) } }
    }

    override fun onError(error: AudioTapError) {
      val payload = NativeAudioTapError(
        code = error.code,
        message = error.message,
        recoverable = error.recoverable,
      )
      errorListeners.values.toList().forEach { runCatching { it(payload) } }
    }
  }

  override val state: String get() = currentStateValue
  override val listenerCount: Double get() = frameListeners.size.toDouble()

  override fun start(format: NativeAudioTapFormat): Promise<Unit> {
    return Promise.async {
      val existing = session
      if (existing != null && existing.isRunning()) return@async

      currentStateValue = "starting"
      stateListeners.values.toList().forEach { runCatching { it("starting") } }

      try {
        val newSession = AudioTapSession(
          context = context,
          sampleRate = format.sampleRate.toInt(),
          channels = format.channels.toInt(),
          bitsPerSample = format.bitsPerSample.toInt(),
          frameDurationMs = format.frameDurationMs.toInt(),
          callbacks = callbacks,
        )
        newSession.start()
        session = newSession
      } catch (err: AudioTapError) {
        session = null
        currentStateValue = "idle"
        stateListeners.values.toList().forEach { runCatching { it("idle") } }
        callbacks.onError(err)
        throw err
      } catch (t: Throwable) {
        session = null
        currentStateValue = "idle"
        stateListeners.values.toList().forEach { runCatching { it("idle") } }
        val wrapped = AudioTapError(
          code = "INTERNAL_ERROR",
          message = t.message ?: t.javaClass.simpleName,
          recoverable = false,
        )
        callbacks.onError(wrapped)
        throw wrapped
      }
    }
  }

  override fun stop(): Promise<Unit> {
    return Promise.async {
      val s = session ?: return@async
      currentStateValue = "stopping"
      stateListeners.values.toList().forEach { runCatching { it("stopping") } }
      runCatching { s.stop() }
      session = null
      currentStateValue = "idle"
      stateListeners.values.toList().forEach { runCatching { it("idle") } }
    }
  }

  override fun release(): Promise<Unit> {
    return Promise.async {
      session?.let { runCatching { it.release() } }
      session = null
      removeAllListeners()
      currentStateValue = "released"
    }
  }

  override fun addFrameListener(listener: (NativeAudioTapFrame) -> Unit): Double {
    val id = nextListenerId.getAndIncrement()
    frameListeners[id] = listener
    return id.toDouble()
  }

  override fun addStateListener(listener: (String) -> Unit): Double {
    val id = nextListenerId.getAndIncrement()
    stateListeners[id] = listener
    return id.toDouble()
  }

  override fun addStatsListener(listener: (NativeAudioTapStats) -> Unit): Double {
    val id = nextListenerId.getAndIncrement()
    statsListeners[id] = listener
    return id.toDouble()
  }

  override fun addErrorListener(listener: (NativeAudioTapError) -> Unit): Double {
    val id = nextListenerId.getAndIncrement()
    errorListeners[id] = listener
    return id.toDouble()
  }

  override fun removeListener(id: Double) {
    val key = id.toInt()
    frameListeners.remove(key)
    stateListeners.remove(key)
    statsListeners.remove(key)
    errorListeners.remove(key)
  }

  override fun removeAllListeners() {
    frameListeners.clear()
    stateListeners.clear()
    statsListeners.clear()
    errorListeners.clear()
  }
}

File: packages/audio-tap/ios/HybridClarionAudioTap.swift
import Foundation
import NitroModules

public typealias AudioTapFrameClosure = (NativeAudioTapFrame) -> Void
public typealias AudioTapStateClosure = (String) -> Void
public typealias AudioTapStatsClosure = (NativeAudioTapStats) -> Void
public typealias AudioTapErrorClosure = (NativeAudioTapError) -> Void

public final class HybridClarionAudioTap: HybridClarionAudioTapSpec {
  private var session: AudioTapSession?
  private var currentStateValue: String = "idle"

  private var nextListenerId: Int = 1
  private var frameListeners: [Int: AudioTapFrameClosure] = [:]
  private var stateListeners: [Int: AudioTapStateClosure] = [:]
  private var statsListeners: [Int: AudioTapStatsClosure] = [:]
  private var errorListeners: [Int: AudioTapErrorClosure] = [:]

  private let listenerLock = NSLock()

  // MARK: - Spec

  public var state: String { currentStateValue }
  public var listenerCount: Double {
    listenerLock.lock(); defer { listenerLock.unlock() }
    return Double(frameListeners.count)
  }

  public func start(format: NativeAudioTapFormat) throws -> Promise<Void> {
    return Promise.async { [weak self] in
      guard let self else { return }
      // Idempotent — re-starting a running tap is a no-op.
      if let existing = self.session, existing.isRunning { return }

      let newSession = AudioTapSession(format: format, callbacks: self)
      self.emit(state: "starting")
      do {
        try await newSession.start()
        self.session = newSession
        self.emit(state: "running")
      } catch let err as AudioTapErr {
        self.session = nil
        self.emit(state: "idle")
        let native = err.toNative()
        self.emit(error: native)
        throw NSError(
          domain: "ClarionAudioTap",
          code: 0,
          userInfo: [NSLocalizedDescriptionKey: "{\"code\":\"\(err.code)\",\"message\":\"\(escape(err.message))\"}"]
        )
      } catch {
        self.session = nil
        self.emit(state: "idle")
        let native = NativeAudioTapError(
          code: "INTERNAL_ERROR",
          message: error.localizedDescription,
          recoverable: false
        )
        self.emit(error: native)
        throw error
      }
    }
  }

  public func stop() throws -> Promise<Void> {
    return Promise.async { [weak self] in
      guard let self, let s = self.session else { return }
      self.emit(state: "stopping")
      await s.stop()
      self.session = nil
      self.emit(state: "idle")
    }
  }

  public func release() throws -> Promise<Void> {
    return Promise.async { [weak self] in
      guard let self else { return }
      if let s = self.session {
        await s.stop()
        self.session = nil
      }
      self.removeAllListenersInternal()
      self.emit(state: "released")
    }
  }

  public func addFrameListener(listener: @escaping (NativeAudioTapFrame) -> Void) throws -> Double {
    return Double(addListener { id in self.frameListeners[id] = listener })
  }

  public func addStateListener(listener: @escaping (String) -> Void) throws -> Double {
    return Double(addListener { id in self.stateListeners[id] = listener })
  }

  public func addStatsListener(listener: @escaping (NativeAudioTapStats) -> Void) throws -> Double {
    return Double(addListener { id in self.statsListeners[id] = listener })
  }

  public func addErrorListener(listener: @escaping (NativeAudioTapError) -> Void) throws -> Double {
    return Double(addListener { id in self.errorListeners[id] = listener })
  }

  public func removeListener(id: Double) throws {
    let key = Int(id)
    listenerLock.lock(); defer { listenerLock.unlock() }
    frameListeners.removeValue(forKey: key)
    stateListeners.removeValue(forKey: key)
    statsListeners.removeValue(forKey: key)
    errorListeners.removeValue(forKey: key)
  }

  public func removeAllListeners() throws {
    removeAllListenersInternal()
  }

  // MARK: - Internal helpers used by AudioTapSession

  internal func emit(frame: NativeAudioTapFrame) {
    let snapshot: [AudioTapFrameClosure]
    listenerLock.lock()
    snapshot = Array(frameListeners.values)
    listenerLock.unlock()
    for closure in snapshot { closure(frame) }
  }

  internal func emit(state: String) {
    currentStateValue = state
    let snapshot: [AudioTapStateClosure]
    listenerLock.lock()
    snapshot = Array(stateListeners.values)
    listenerLock.unlock()
    for closure in snapshot { closure(state) }
  }

  internal func emit(stats: NativeAudioTapStats) {
    let snapshot: [AudioTapStatsClosure]
    listenerLock.lock()
    snapshot = Array(statsListeners.values)
    listenerLock.unlock()
    for closure in snapshot { closure(stats) }
  }

  internal func emit(error: NativeAudioTapError) {
    let snapshot: [AudioTapErrorClosure]
    listenerLock.lock()
    snapshot = Array(errorListeners.values)
    listenerLock.unlock()
    for closure in snapshot { closure(error) }
  }

  // MARK: - Private

  private func addListener(_ register: (Int) -> Void) -> Int {
    listenerLock.lock()
    let id = nextListenerId
    nextListenerId += 1
    register(id)
    listenerLock.unlock()
    return id
  }

  private func removeAllListenersInternal() {
    listenerLock.lock()
    frameListeners.removeAll()
    stateListeners.removeAll()
    statsListeners.removeAll()
    errorListeners.removeAll()
    listenerLock.unlock()
  }
}

/// Minimal JSON escape so the JS-side `tryParseStructured` can recover the
/// native code/message from an NSError.localizedDescription.
private func escape(_ s: String) -> String {
  return s.replacingOccurrences(of: "\\", with: "\\\\")
    .replacingOccurrences(of: "\"", with: "\\\"")
    .replacingOccurrences(of: "\n", with: "\\n")
}
